codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt

```diff
@@ -572,42 +573,70 @@ private class ServerHttpProtocolImplGenerator(
         if (pathBindings.isEmpty()) {
             return
         }
-        val pattern = StringBuilder()
         val httpTrait = httpBindingResolver.httpTrait(operationShape)
-        httpTrait.uri.segments.forEach {
-            pattern.append("/")
-            if (it.isLabel) {
-                pattern.append("(?P<${it.content}>")
-                if (it.isGreedyLabel) {
-                    pattern.append(".+")
-                } else {
-                    pattern.append("[^/]+")
-                }
-                pattern.append(")")
-            } else {
-                pattern.append(it.content)
-            }
-        }
         with(writer) {
             rustTemplate(
                 """
-                #{LazyStatic}::lazy_static! {
-                    static ref RE: #{Regex}::Regex = #{Regex}::Regex::new("$pattern").unwrap();
-                }
+                let input_string = request.uri().path();
                 """.trimIndent(),
                 *codegenScope,
             )
-            rustBlock("if let Some(captures) = RE.captures(request.uri().path())") {
-                pathBindings.forEach {
-                    val deserializer = generateParsePercentEncodedStrFn(it)
+            for (index in httpTrait.uri.segments.indices) {
+                val segment = httpTrait.uri.segments[index]
+                val name = segment.content
+                val binding = pathBindings.find { it.memberName == name }
+                val inputStringPrefix = if (index == httpTrait.uri.segments.size - 1) { "_" } else { "" }
+                val bindingPrefix = if (!(binding != null && segment.isLabel)) { "_" } else { "" }
+                rustTemplate(
+                    """
+                    let (input_string, _tag) = #{Nom}::bytes::complete::tag::<_, &str, #{Nom}::error::Error<&str>>("/")(input_string).unwrap();
+                    """.trimIndent(),
+                    *codegenScope
+                )
+                if (segment.isGreedyLabel) {
+                    val end = httpTrait.uri.segments.slice((index + 1) until httpTrait.uri.segments.size).joinToString(prefix = "/", separator = "/")
+                    if (end.length > 1)
+                        rustTemplate(
+                            """
+                            let (${bindingPrefix}m, ${inputStringPrefix}input_string) = match input_string.rfind(${end.dq()}) {
+                                Some(n) => (&input_string[..n], &input_string[n..]),
+                                None => (input_string, "")
```

Comment:
* Here we're extracting to the first pair component, which is the reverse of nom's API that we use in the other places, so it's confusing.
* If the suffix is not found, we're assigning the field bound to the greedy label to the entire `input_string` when we should error out.